pop-api/examples/fungibles/tests.rs

```diff
@@ -0,0 +1,828 @@
+}
+
+/// 1. PSP-22 Interface:
+/// - total_supply
+/// - balance_of
+/// - allowance
+/// - transfer
+/// - transfer_from
+/// - approve
+/// - increase_allowance
+/// - decrease_allowance
+
+#[drink::test(sandbox = Sandbox)]
+fn total_supply_works(mut session: Session) {
+	let _ = env_logger::try_init();
+	// Deploy a new contract.
+	deploy(&mut session, CONTRACT.clone(), "new", vec![TOKEN.to_string(), MIN_BALANCE.to_string()])
+		.unwrap();
+
+	// No tokens in circulation.
+	assert_eq!(total_supply(&mut session), 0);
+	assert_eq!(total_supply(&mut session), session.sandbox().total_supply(&TOKEN));
+
+	// Tokens in circulation.
+	assert_ok!(session.sandbox().mint_into(&TOKEN, &ALICE, AMOUNT));
+	assert_ok!(session.sandbox().mint_into(&TOKEN, &BOB, AMOUNT * 2));
+	assert_eq!(total_supply(&mut session), AMOUNT * 3);
+	assert_eq!(total_supply(&mut session), session.sandbox().total_supply(&TOKEN));
+}
+
+#[drink::test(sandbox = Sandbox)]
+fn balance_of_works(mut session: Session) {
+	let _ = env_logger::try_init();
+	// Deploy a new contract.
+	deploy(&mut session, CONTRACT.clone(), "new", vec![TOKEN.to_string(), MIN_BALANCE.to_string()])
+		.unwrap();
+
+	// No tokens in circulation.
+	assert_eq!(balance_of(&mut session, ALICE), 0);
+	assert_eq!(balance_of(&mut session, ALICE), session.sandbox().balance_of(&TOKEN, &ALICE));
+
+	// Tokens in circulation.
+	assert_ok!(session.sandbox().mint_into(&TOKEN, &ALICE, AMOUNT));
+	assert_eq!(balance_of(&mut session, ALICE), AMOUNT);
+	assert_eq!(balance_of(&mut session, ALICE), session.sandbox().balance_of(&TOKEN, &ALICE));
+}
+
+#[drink::test(sandbox = Sandbox)]
+fn allowance_works(mut session: Session) {
+	let _ = env_logger::try_init();
+	// Deploy a new contract.
+	let contract = deploy(
+		&mut session,
+		CONTRACT.clone(),
+		"new",
+		vec![TOKEN.to_string(), MIN_BALANCE.to_string()],
+	)
+	.unwrap();
+
+	// No tokens in circulation.
+	assert_eq!(allowance(&mut session, contract.clone(), ALICE), 0);
+
+	// Tokens in circulation.
+	assert_ok!(session.sandbox().mint_into(&TOKEN, &contract.clone(), AMOUNT));
+	assert_ok!(session.sandbox().approve(&TOKEN, &contract.clone(), &ALICE, AMOUNT / 2));
+	assert_eq!(allowance(&mut session, contract, ALICE), AMOUNT / 2);
+}
+
+#[drink::test(sandbox = Sandbox)]
+fn transfer_works(mut session: Session) {
+	let _ = env_logger::try_init();
+	// Deploy a new contract.
+	let contract = deploy(
+		&mut session,
+		CONTRACT.clone(),
+		"new",
+		vec![TOKEN.to_string(), MIN_BALANCE.to_string()],
+	)
+	.unwrap();
+	session.set_actor(contract.clone());
+	// `pallet-assets` returns `NoAccount` error.
+	assert_eq!(
+		transfer(&mut session, ALICE, AMOUNT / 4),
+		Err(PSP22Error::Custom(
+			vec_u8_to_u32(Module { index: 52, error: [1, 0] }.encode()).unwrap().to_string()
+		))
+	);
+
+	// Mint tokens.
+	assert_ok!(session.sandbox().mint_into(&TOKEN, &contract.clone(), AMOUNT));
+	assert_ok!(session.sandbox().mint_into(&TOKEN, &BOB, AMOUNT));
+
+	// Successfully transfer.
```

Comment:
Reason why I move it to the top coming from a comment that we want to test based on the logic order in the standard.

Reply:
Aa oke, could you tell me where this is specified in the standard or the comment you are referring to?

Reply:
Actually, I followed the description of the trait not the function logic. It makes sense to bring the success cast to the bottom because this is what I did in the beginning.